cli/disputer.go

```diff
@@ -0,0 +1,359 @@
+		defer ticker.Stop()
+		// TODO: Don't start unless fully synced (check syncstatus)
+		for {
+			select {
+			case notif, ok := <-headChanges:
+				if !ok {
+					return xerrors.Errorf("head change channel errored")
+				}
+				for _, val := range notif {
+					switch val.Type {
+					case store.HCApply:
+						// TODO: Confirm that we're never notified to "apply" a null tipset
+
+						if val.Val.Height() < miner3.WPoStChallengeWindow {
+							// it's too early in the chain to challenge any posts
+							continue
+						}
+
+						// TODO: If the newly applied tipset is built on top of n null tipset(s), we should also check the tipsets WPoStChallengeWindow+[1...n]  epochs ago
+
+						// Since we're waiting WPoStChallengeWindow epochs already, we don't have to worry about confidence of the message -- it's already pretty deep onchain
+						ts, err := api.ChainGetTipSetByHeight(ctx, val.Val.Height()-miner3.WPoStChallengeWindow, val.Val.Key())
+						if err != nil {
+							return xerrors.Errorf("failed to get tipset 60 epochs ago: %w", err)
+						}
+
+						if ts.IsNull() {
+							continue
+						}
+
+						// TODO: Build a cache of tipsets we've already inspected and continue here if appropriate
+
+						pmsgs, err := api.ChainGetParentMessages(ctx, ts.Cids()[0])
+						if err != nil {
+							return xerrors.Errorf("failed to get parent messages: %w", err)
+						}
+
+						wpmsgs, err := getSubmitWindowedPoSts(ctx, api, pmsgs, ts.Key())
+						if err != nil {
+							return xerrors.Errorf("failed to get post messages: %w", err)
+						}
+
+						dpmsgs, err := makeDisputeWindowedPosts(ctx, api, wpmsgs, fromAddr)
+						if err != nil {
+							return xerrors.Errorf("failed to check for disputes: %w", err)
+						}
+
+						for _, dpmsg := range dpmsgs {
+							fmt.Println("disputing a post at height ", ts.Height())
+							_, err := api.MpoolPushMessage(ctx, dpmsg, nil)
+							if err != nil {
+								// TODO: This might not be worth returning on...
+								return xerrors.Errorf("failed to dispute post message: %w", err)
+							}
+
+							// TODO: Track / report on message landing on chain?
+						}
+					case store.HCRevert:
+						// do nothing
+					default:
+						return xerrors.Errorf("unexpected head change type %s", val.Type)
+					}
+				}
+			case <-ticker.C:
+				fmt.Print("Running health check: ")
+
+				cctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+
+				if _, err := api.ID(cctx); err != nil {
+					fmt.Println("Health check failed")
+					cancel()
+					return nil
+				}
+
+				cancel()
+
+				fmt.Println("Node online")
+			case <-ctx.Done():
+				return nil
+			}
+		}
+	},
+}
+
+// for a list of messages and the tsk in which they were executed, returns the _successful_ SubmitWindowedPoSt msgs among them
+func getSubmitWindowedPoSts(ctx context.Context, fullApi api.FullNode, msgs []api.Message, tsk types.TipSetKey) ([]api.Message, error) {
+	posts := make([]api.Message, 0)
+	for _, msg := range msgs {
```

Comment:
Could run this in parallel (probably using `lib/parmap.Par(50, msgs, func(api.Message){...})`)